src/coreclr/src/jit/gentree.cpp

```diff
@@ -12498,6 +12498,26 @@ GenTree* Compiler::gtFoldTypeCompare(GenTree* tree)
         objOp = opOther->AsCall()->gtCallThisArg->GetNode();
     }
 
+    bool                 pIsExact   = false;
+    bool                 pIsNonNull = false;
+    CORINFO_CLASS_HANDLE objCls     = gtGetClassHandle(objOp, &pIsExact, &pIsNonNull);
+
+    // if both classes are "final" (e.g. System.String[]) we can replace the comparison
+    // with `true/false` + null check.
+    if ((objCls != NO_CLASS_HANDLE) &&
+        // double check if it's final via impIsClassExact
+        impIsClassExact(objCls) && impIsClassExact(clsHnd))
```

Comment:
I think this can be simply `isExact || impIsClassExact(objCls)`.

We know exactly which type `clsHnd` refers to.